extensions/resteasy-classic/resteasy-client/deployment/src/test/java/io/quarkus/restclient/configuration/RestClientRandomPortTest.java

```diff
@@ -21,7 +21,7 @@ class RestClientRandomPortTest {
                     .addClasses(EchoResource.class, EchoClient.class))
             .overrideRuntimeConfigKey("quarkus.http.port", "0")
             .overrideRuntimeConfigKey("quarkus.http.test-port", "0")
-            .overrideRuntimeConfigKey("quarkus.rest-client.EchoClient.url", "http://localhost:${quarkus.http.port}");
+            .overrideRuntimeConfigKey("quarkus.rest-client.EchoClient.url", "quarkus://vertx-http");
```

Comment:
We should not use URLs in this way. The URL specifies the protocol to use, not the address. We could potentially have a special syntax for referencing a service address within a configuration key, alongside an "aware" parser/property type that knows how to parse it. We'd need one for URL configs, and a separate one (or maybe two) for socket address/port number configs.

Reply:
The reason it's bad, by the way, just in this case for example we don't know whether it's `http` or `https`. We could theoretically glean that information from the service, but that's not 100% foolproof either because there could conceivably be URL schemes which are layered combinations of protocols, where the server would (at best) only be aware of the outer protocol. But that's just one example; any time you have a mismatch like this, it will be accompanied by a variety of subtle and hard-to-fix problems.

Reply:
In this case `quarkus://vertx-http` is not an `URL` at all. Is just the key to lookup the service. It could has been `foo.bar`, as long as the service would register itselft as `foo.bar`. The service lookup would then return the proper `URL`.

I admit it is confusing because on how I choose the key name and the configuration name (which is current name).

Reply:
Right, I got that. But the problem is that the service is identified on the client differently than on the server. We lose information that would normally be found only in the client URL scheme.

Reply:
This was just a way to make a short alias and then for us to discuss. And doesnt' have to be this way.

I'm not expecting to ask users to go and change all their configuration for service names, so setting the standard url as `http://localhost:${quarkus.http.port}` would still need to be valid and we would need to rewrite the port (in case of randoms to use the right port).

Or we can just drop the names / alias completely and do only the rewriting of the port when required. Or do you want to make a hard break on that?

Reply:
It's a little complex because services could be defined by host+port, by *just* port, by socket address (`123.45.67.89:1234`), etc. And, clients might define their access to that service by URL, by socket address, or by host+port, or some other creative scheme.

I think the abstraction of a "service" or "network service" is a good start. We can strictly define them as a "named" socket address. This is unambiguous (and actually corresponds reasonably well with [what WildFly does](https://docs.jboss.org/author/display/WFLY/Interfaces%20and%20ports.html)).

If a client connects by socket address, then we would need a way to say "socket address *or* service", with a corresponding class that works with `Converter` that can give the correct `SocketAddress` when given access to the registry of running services.

If a client connects by URL, then we would need a way represent a URL with either a "normal" host/port part or else some token which signifies a local service should be used. This would be an extension relative to valid URL syntax so we cannot use `java.net.URL/URI` to represent these; rather we'd have a `Converter`-friendly class which can hold either a plain URL or else something that can resolve to a URL when given access to the registry of running services.

We could use a uniform syntax for these cases e.g. `@{serviceName}` or `https://@{serviceName}/my/path`, however if we do so, it should be made very clear that this syntax is not resolved by the configuration layer like properties are, but rather by the property's `Converter` itself.

Reply:
Except it should not be named service, I prefer David's suggestion.